### ISSUE TITLE
Scroll preview

### ISSUE THREAD

**Issue by jdfa** [gokcehan/lf#925]
For cases when previewed file contains few pages it would make sense to introduce command for scrolling preview up and down.
It would be something similar to what fzf has.
That would also require indicator of current/total line.

**Comment by ilyagr** [comment]
This seems unlikely to happen. You can look at https://github.com/gokcehan/lf/issues/610 and https://github.com/gokcehan/lf/issues/568.

**Comment by jdfa** [comment]
Oh, I searched only through opened issues for this.
I still don't get all reasoning behind Pager argumentation. In my understanding, preview content is pretty similar to folder content in terms of scrolling: some multiline data which is fully generated on first call.

**Comment by MinmoTech** [comment]
With ranger you can scroll the preview with the mouse, it would be awesome if lf could do the same!

To iterate why it's not the same as opening the file with a pager, for some binary files I use text output as a preview (ffprobe or mediainfo for videos) and I'd love to be able to scroll that!

(pinging @gokcehan as I think this is a relevant point of discussion still and this is the most informative open issue about the topic)

I do understand that it's incompatible with the preview caching, so that might still be a strong argument against preview scrolling.

**Comment by gokcehan** [comment]
@MinmoTech I'm really tired of saying no to feature requests, so if someone figures out a way to make this work, feel free to send it as a patch and I will let our contributors decide whether or not the patch should be accepted. The only thing I require is that it should somehow not disable the preview optimization (i.e. only read the first portion of the file) for users that are not interested in this behavior. I'm not sure if it is possible or not.

**Comment by unphased** [comment]
i finally worked out how to get lf set up to do some nice things (xxd for binary files, jq to prettify json, all done via simple `file` logic in a shell script) but the glaring limitation with preview now is not being able to scroll! 

Sure i can open a file in neovim to gain scrolling but my nvim config is a bit heavy at the moment and that's too slow. 

I am curious if you could succinctly describe a little more what the preview optimization does. Does it come into play if you define a previewer?

**Comment by joelim-work** [comment]
`lf` stores directory contents in memory so it is possible to scroll through them. Although it sounds similar, storing file contents is not the same because file contents can be very large (i.e. several GB) which can lead to out-of-memory issues.

That being said, it's still possible to achieve this via external configuration by using `tail` in the `previewer` script to read the file starting from the scroll offset. See the [tip](https://github.com/gokcehan/lf/wiki/Tips#scrollable-previews) in the wiki for how to configure this.

**Comment by unphased** [comment]
I'll update with what i use now for lf to open files since i like to use less rather than vim/nvim to view most files (for speed): 

```
cmd open ${{
    set -f
    filesize=$(stat -f%z "$f")
    case "$f" in
        *.json)
            jq . "$f" | bat --paging=always --file-name "$f"
            ;;
        *.log|*.log.[0-9]*)
            bat --paging=always "$f"
            ;;
        *)
            if [ "$filesize" -gt 2097152 ]; then
                bat --paging=always "$f"
            else
                # echo "Launching: nvim $f" >> ~/.lf_launches.log
                nvim "$f"
            fi
            ;;
    esac
}}
```

Thank you for sharing the interesting hack method to build a scrollable preview. Looks like it will be easy to incorporate into my preview script I use.

**Comment by unphased** [comment]
I got the suggested preview system working and it can scroll, but flickers like mad and has significant input latency. I'm disappointed. This is why hacks relying on IPC with no regard for process launch overhead don't pan out. I can't believe someone looked at how this behaves and thinks it's an adequate solution.

**Comment by joelim-work** [comment]
I agree that the solution is not exactly performant, as processes are launched every time the user wants to scroll, but this is the price of combining minimalistic software in order to 'build' a feature.

Regarding the idea of reading files in chunks on-demand, what you're describing is pretty much implementing a pager directly in `lf`, and this was marked as a non-feature from the very start, see https://github.com/gokcehan/lf/issues/610#issuecomment-1030068462. I'm not saying that it is impossible to implement, that is far from the case and you are of course welcome to try. However the reality is that if this feature is implemented directly as a builtin, then it will have to be maintained, and users will invariably request other features like `top`/`bottom`, searching, bookmarking, etc., and at some point it becomes less clear of whether `lf` is supposed to be a file manager or file viewer.

The solution I provided is intended to be a way to quickly scroll through the contents of the file just to see if it 'looks right'. If I am interested in viewing this file in detail, then I could press a keybinding to open it in `less`/`vim` or some other appropriate program.

**Comment by unphased** [comment]
I mostly vibecoded something that works pretty well for scrolling preview, it's a shell script used as the previewer script. It manages cache files that are `tail`ed at different offsets to render scrolling quickly enough, and it uses modified time to invalidate the files in this cache. Some of the biggest remaining drawbacks:

- it still flickers when scrolling, i do wonder if porting it to C++ could make it fast enough not to flicker, but the entire approach is probably too flawed to go much further. still, the flicker is only for one frame and not honestly that bad...
- no ability to know the height of the lf preview window. I wonder if this is available in lf config for a command to pass along. If so I can at least implement a bind that is likely to be more useful than any scrolling, which is to snap the preview to view the bottom of a (e.g. log) file. 

Even better improvement on the last bullet would be if I can establish a piece of state so i can toggle the preview always showing the end of the log file. Possibly have this logic sticky based on dir and file extension.

**Comment by joelim-work** [comment]
If it helps you, the height of the preview window is passed to the `previewer` script: https://github.com/gokcehan/lf/blob/master/doc.md#previewer-string-default--not-filtered-if-empty

> Set the path of a previewer file to filter the content of regular files for previewing. The file should be executable. The following arguments are passed to the file, (1) current file name, (2) width, (3) height, (4) horizontal position, and (5) vertical position of preview pane respectively. SIGPIPE signal is sent when enough lines are read. If the previewer returns a non-zero exit code, then the preview cache for the given file is disabled. This means that if the file is selected in the future, the previewer is called once again. Preview filtering is disabled and files are displayed as they are when the value of this option is left empty.

**Comment by unphased** [comment]
One of the issues I'm running into is that it's impossible to communicate the length of the preview output back to `lf` in order to implement behavior to clamp the `lf_user_preview_offset` value to the number of lines contained in the preview.

**Comment by joelim-work** [comment]
> Somewhat equivalently it's not possible to delegate the assignment of that value to the previewer script...

You can use `lf -remote` to set values even in the `previewer` script, so long as you specify the id:

```sh
lf -remote "send $id set user_preview_lines $(wc -l "$1" | cut -d ' ' -f 1)"
```

**Comment by unphased** [comment]
Thank you. that looks useful, but the blueprint I came up with has panned out. I now have full desired functionality aside from the delay of launching the script causing a quick blank preview page blinking. It would be really great if `lf` could just not paint the preview's content until it is received, then the single frame render delay would be imperceptible and we wouldn't need to deal with the blinking. I'll prep it into a gist.

**Comment by unphased** [comment]
not excessively tested yet but already fast and usable and has built in logging. https://gist.github.com/unphased/f078a5d4d974b141ccd43c224a5dd5af

**Comment by voronind-com** [comment]
I actually created a map key that pipes preview to the pager. Replace `@previewer@` with your preview tool. This way I can press `I` to see the full preview when needed.

```text
map I full_preview
cmd full_preview ${{ @previewer@ $f | $PAGER }}
```
